Limit type changes

orangejenny opened the pull request:
http://manage.dimagi.com/default.asp?67177

Left the alert alone to catch any programmatic attempts to change a select with children to a non-select.

millerdev commented on src/core.js:
```diff
@@ -1228,6 +1228,10 @@ define([
+            var currentMug = _this.getCurrentlySelectedMug();
+            if (e.mug && e.mug.parentMug && e.mug.parentMug === currentMug) {
+                _this.displayMugProperties(currentMug);
+            }
```
Was there a particular reason this was put at the beginning of this callback? Seems like `selectSomethingOrHideProperties` will possibly call `displayMugProperties` again. Would be good to not display properties more than once.

millerdev replied:
Not only `selectSomethingOrHideProperties`, but anything that calls `jstree("select_node", ...)` will likely call `displayMugProperties`.

orangejenny replied:
This isn't necessary, I did it for symmetry with the code below in case we ever programmatically remove children. I'm fine with removing it.

orangejenny replied:
...actually, the one place we programmatically remove children is in the test I added in this PR. Leaving this code in here unless you strongly object.

millerdev replied:
It probably won't matter, but we should be mindful of this if we start to notice the properties flashing or other general sluggishness on creating/moving/removing questions.

millerdev commented on src/core.js:
```diff
@@ -1242,6 +1246,10 @@ define([
+            }
             if (!e.isInternal) {
                 _this.jstree("deselect_all", true)
                      .jstree('select_node', e.mug.ufid);
```
Same

orangejenny replied:
This one's necessary because when the two child options are auto-added to a new select, that should cause the type dropdown to change, but those events are internal, so the select_node doesn't get triggered.

millerdev commented on src/core.js:
```diff
@@ -1655,14 +1663,17 @@ define([
                     (!q.limitTypeChangeTo || 
-                     q.limitTypeChangeTo.indexOf(currentType) !== -1))
+                     q.limitTypeChangeTo.indexOf(currentType) !== -1) &&
+                    (!hasChildren || !isSelect(mug.__className) || isSelect(typeName)))
```
I know I was the one to refactor similar code over [here](https://github.com/dimagi/Vellum/blob/limit-type-changes/src/mugs.js#L1080), but since then I have started to delegate this kind of decision to the mug so we don't have to hard-code type names in places like this. This also makes it easier for plugins like itemsets to write custom mug types that need to have a say in this decision. What do you think of adding a new mug option like this:

```js
if (q.isTypeChangeable && currentType !== typeName &&
    (!q.limitTypeChangeTo ||
     q.limitTypeChangeTo.indexOf(currentType) !== -1) &&
    mug.options.canChangeType(mug, typeName))
{
    ...
}
```
`defaultOptions` in mugs.js will have this:
```js
/**
 * Decide if this mug can have its type changed to typeName
 *
 * The value of `limitTypeChangeTo` takes precedence over this option.
 *
 * @param mug - the mug that would change type.
 * @param typeName - the name of the new type for mug (e.g., MSelect).
 * @returns - true if the mug can change to type, otherwise false.
 */
canChangeType: function (mug, typeName) { return true; },
```

Then MSelect, and by inheritance Select also, can provide a custom value for this option that checks if the mug has children, etc.

orangejenny replied:
I like that much better, will implement it.

orangejenny commented:
@millerdev 

Addressed comments. Implemented canChangeType as a function that returns an error message on failure. It's a little non-standard, but it lets me move the actual error message into mugs.js. Also eliminated limitTypeChangeTo, because having three different ways to specify type changing seemed like a bit much.

millerdev commented:
Looks good. Waiting for build.
